program/src/processor.rs

```diff
@@ -325,28 +369,29 @@ impl Processor {
     }
 
     fn process_delegate(accounts: &[AccountInfo]) -> ProgramResult {
+        let signers = collect_signers(accounts);
         let account_info_iter = &mut accounts.iter();
+
+        // native asserts: 5 accounts (2 sysvars + stake config)
         let stake_account_info = next_account_info(account_info_iter)?;
         let vote_account_info = next_account_info(account_info_iter)?;
         let clock_info = next_account_info(account_info_iter)?;
-        let clock = &Clock::from_account_info(clock_info)?;
         let _stake_history_info = next_account_info(account_info_iter)?;
         let _stake_config_info = next_account_info(account_info_iter)?;
-        let stake_authority_info = next_account_info(account_info_iter)?;
 
-        let stake_history = &StakeHistorySysvar(clock.epoch);
+        // other accounts
+        // let _stake_authority_info = next_account_info(account_info_iter);
```

Comment:
im going to do a pr after this one to implement a version of `next_account_info()` which skips sysvars and stake config. this should be a perf improvement for 100% of transactions, once people omit the accounts, because getting sysvars by syscall is cheaper than building the `AccountInfo` in `entrypoint!`. once this program is live we should change the instruction builders to omit sysvars and backport all the way to 2.0 no matter what version is on mainnet

once i replace `next_account_info()`, this code becomes:
* return error if fewer than three accounts
* collect signers
* create accounts iter
* unwrap two accounts

then if we go forward with breaking interface changes, we would unwrap a third account, the stake authority, and it will do the right thing regardless of whether stake authority is third or sixth

i point this out here (rather than for all the other instructions where authority is technically optional) because the monorepo instruction builder for `Delegate` still includes stake config for "backwards compatibility," even tho its not technically necessary because stake authority is not asserted to exist, but we should leave it as-is to avoid making this even more complicated

in other words no one using our instruction builders would ever place stake authority fifth

Reply:
Just be careful -- skipping the sysvar accounts is a breaking change between this and native, since at the very least we should check the sysvar account address.

Unless there's a security reason, we should not break the instruction interface, and instead introduce new instructions, but have the old instruction creators create the new instructions -- ie `fn delegate()` creates a `DelegateNew` instruction.

There's a lot of staking tooling / programs that rely on the current instructions, so we can't break them. After we introduce the new instructions, we can give people a couple of years before removing them.

Reply:
what i mean is write a `next_account_info()` that does something like this:
```rust
fn hana_next_account(account_iter: &mut _) -> ProgramResult<AccountInfo> {
  loop {
    let account = next_account_info(account_iter);
    match &account {
      Ok(a) if is_sysvar_or_stake_config(a) => continue,
      _ => return account,
    }
  }
}
```

in other words, we dont change the positionality of any accounts, we just advance past ones we dont care about. that way we can support old and new instructions in a non-breaking way (unless you are thinking of a different kind of breaking change)

im also not opposed to just making a second brand new interface that does everything right and deprecating the old one tho

Reply:
Ah ok, that makes sense! Sorry, I was worried that you were suggesting breaking changes on the old instructions. 

Both approaches are fine by me, we can see what it looks like when we get there. I did really like how you implemented move-stake / move-lamports, so a brand new interface might be the better option.